blueocean-dashboard/src/main/java/io/jenkins/blueocean/preload/PipelineRunsStatePreloader.java

```diff
@@ -0,0 +1,116 @@
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Preload pipeline runs onto the page if the requested page is a pipeline runs page.
+ *
+ * @author <a href="mailto:tom.fennelly@gmail.com">tom.fennelly@gmail.com</a>
+ */
+@Extension
+public class PipelineRunsStatePreloader extends RESTFetchPreloader {
+
+    private static final Logger LOGGER = Logger.getLogger(PipelineRunsStatePreloader.class.getName());
+
+    private static final int DEFAULT_LIMIT = 26;
+
+    @Override
+    protected FetchData getFetchData(@Nonnull BlueUrlTokenizer blueUrl) {
+        BluePipeline pipeline = getPipeline(blueUrl);
+
+        if (pipeline != null) {
+            // It's a pipeline page. Let's prefetch the pipeline runs and add them to the page,
+            // saving the frontend the overhead of requesting them.
+
+            BlueRunContainer runsContainer = pipeline.getRuns();
+            Iterator<BlueRun> runsIterator = runsContainer.iterator(0, DEFAULT_LIMIT);
+            JSONArray runs = new JSONArray();
+
+            while(runsIterator.hasNext()) {
+                BlueRun blueRun = runsIterator.next();
+                try {
+                    runs.add(JSONObject.fromObject(ModelObjectSerializer.toJson(blueRun)));
+                } catch (IOException e) {
+                    LOGGER.log(Level.FINE, String.format("Unable to preload runs for Job '%s'. Run serialization error.", pipeline.getFullName()), e);
+                    return null;
+                }
+            }
+
+            return new FetchData(
+                pipeline.getActivities().getLink().getHref() + "?start=0&limit=" + DEFAULT_LIMIT,
+                runs.toString());
+        }
+
+        // Don't preload any data on the page.
+        return null;
+    }
+
+    private BluePipeline getPipeline(BlueUrlTokenizer blueUrl) {
+        if (addPipelineRuns(blueUrl)) {
+            Jenkins jenkins = Jenkins.getInstance();
+            String pipelineFullName = blueUrl.getPart(BlueUrlTokenizer.UrlPart.PIPELINE);
+
+            try {
+                Item pipelineJob = jenkins.getItemByFullName(pipelineFullName);
+                return (BluePipeline) BluePipelineFactory.resolve(pipelineJob);
+            } catch (Exception e) {
+                LOGGER.log(Level.FINE, String.format("Unable to find Job named '%s'.", pipelineFullName), e);
+                return null;
+            }
+        }
+
+        return null;
+    }
+
+    private boolean addPipelineRuns(@Nonnull BlueUrlTokenizer blueUrl) {
+        if (blueUrl.lastPartIs(BlueUrlTokenizer.UrlPart.PIPELINE)) {
+            // e.g. /blue/organizations/jenkins/f1%2Ff3%20with%20spaces%2Ff3%20pipeline/
+            return true;
+        } else if (blueUrl.lastPartIs(BlueUrlTokenizer.UrlPart.PIPELINE_TAB, "activity")) {
+            // e.g. /blue/organizations/jenkins/f1%2Ff3%20with%20spaces%2Ff3%20pipeline/activity/
+            return true;
+        }
+
+        return false;
+    }
```

Comment:
You are already doing something similar to the `shouldPreload` method here, so maybe that brings some weight to the idea.